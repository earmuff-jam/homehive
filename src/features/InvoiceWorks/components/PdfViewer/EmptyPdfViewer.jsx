import { Typography } from "@mui/material";
import EmptyComponent from "common/EmptyComponent";

export default function EmptyPdfViewer({ handleNavigate }) {
  return (
    <EmptyComponent
      title="Sorry, no invoice found to display"
      caption="Create new invoice form"
      sxProps={{ textTransform: "initial" }}
    >
      <Typography
        component={"span"}
        variant="caption"
        color="primary"
        sx={{ cursor: "pointer" }}
        onClick={handleNavigate}
      >
        {" "}
        here.
      </Typography>
    </EmptyComponent>
  );
}
