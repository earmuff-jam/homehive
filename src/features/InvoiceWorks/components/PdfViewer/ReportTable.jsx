import React from "react";

import {
  Box,
  Paper,
  Stack,
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
  Typography,
} from "@mui/material";

export default function ReportTable({
  rows = [],
  taxRate = 0,
  showWatermark,
  invoiceStatus,
  invoiceTitle = "Invoice Details",
}) {
  const numberFormatter = (number) => Number(number).toFixed(2);

  const subtotal = rows
    .map(({ price }) => parseFloat(price) || 0)
    .reduce((sum, i) => sum + i, 0);

  const paymentRecieved = rows
    .map(({ payment }) => parseFloat(payment) || 0)
    .reduce((sum, i) => sum + i, 0);

  const invoiceSubtotal = subtotal - paymentRecieved;
  const formattedTax = parseFloat(taxRate) || 0;
  const invoiceTaxes = (formattedTax / 100) * invoiceSubtotal;
  const invoiceTotal = invoiceSubtotal + invoiceTaxes;

  return (
    <TableContainer component={Paper}>
      <Table sx={{ minWidth: 700 }} aria-label="spanning table">
        <TableHead>
          <TableRow>
            <TableCell align="center" colSpan={5} sx={{ fontWeight: "bold" }}>
              {invoiceTitle}
            </TableCell>
            <TableCell align="right" sx={{ fontWeight: "bold" }}>
              Price (USD)
            </TableCell>
          </TableRow>
          <TableRow>
            <TableCell sx={{ fontWeight: "bold" }}>Category</TableCell>
            <TableCell sx={{ fontWeight: "bold" }}>Description</TableCell>
            <TableCell align="right" sx={{ fontWeight: "bold" }}>
              Qty.
            </TableCell>
            <TableCell align="right" sx={{ fontWeight: "bold" }}>
              Cost
            </TableCell>
            <TableCell align="right" sx={{ fontWeight: "bold" }}>
              Payment Received
            </TableCell>
            <TableCell align="right" sx={{ fontWeight: "bold" }}>
              Balance Due
            </TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {rows.map((row, index) => (
            <TableRow key={index}>
              <TableCell>
                <Typography variant="subtitle">
                  {row?.category || ""}
                </Typography>
              </TableCell>
              <TableCell>
                <Stack direction="row" spacing={1} alignItems="flex-end">
                  <Typography variant="subtitle">
                    {row.descpription || ""}
                  </Typography>
                  <Typography variant="caption">
                    <sub>{row.caption || ""}</sub>
                  </Typography>
                </Stack>
              </TableCell>
              <TableCell align="right">{row.quantity}</TableCell>
              <TableCell align="right">{numberFormatter(row.price)}</TableCell>
              <TableCell align="right">
                {numberFormatter(row.payment)}
              </TableCell>
              <TableCell align="right">
                {numberFormatter(
                  (parseFloat(row.price) || 0) - (parseFloat(row.payment) || 0),
                )}
              </TableCell>
            </TableRow>
          ))}
          <TableRow>
            <TableCell rowSpan={3} />
            <TableCell colSpan={4} sx={{ fontWeight: "bold" }}>
              Subtotal
            </TableCell>
            <TableCell align="right" sx={{ fontWeight: "bold" }}>
              {numberFormatter(invoiceSubtotal)}
            </TableCell>
          </TableRow>

          <TableRow>
            <TableCell sx={{ fontWeight: "bold" }}>Tax</TableCell>
            <TableCell colSpan={3} align="right" sx={{ fontWeight: "bold" }}>
              {formattedTax}%
            </TableCell>
            <TableCell align="right" sx={{ fontWeight: "bold" }}>
              {numberFormatter(invoiceTaxes)}
            </TableCell>
          </TableRow>

          <TableRow>
            <TableCell colSpan={4} sx={{ fontWeight: "bold" }}>
              <Stack direction="row" spacing={10}>
                <Typography>Total</Typography>
                <Box component="span">
                  {invoiceStatus?.display && (
                    <Typography
                      color="error.light"
                      variant="caption"
                      className={!showWatermark && "no-print"} // print only when allowed
                      sx={{
                        textTransform: "uppercase",
                        fontSize: "2rem",
                        padding: "0.3rem 0.5rem",
                        textAlign: "center",
                        border: "0.3rem solid",
                      }}
                    >
                      {invoiceStatus?.label}
                    </Typography>
                  )}
                </Box>
              </Stack>
            </TableCell>
            <TableCell align="right" sx={{ fontWeight: "bold" }}>
              {numberFormatter(invoiceTotal)}
            </TableCell>
          </TableRow>
        </TableBody>
      </Table>
    </TableContainer>
  );
}
