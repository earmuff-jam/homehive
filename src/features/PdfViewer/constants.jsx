import dayjs from "dayjs";

/**
 * Invoice Category Options
 *
 * used to build out the autocomplete component in edit / view invoice
 */
export const InvoiceCategoryOptions = [
  { label: "Products", value: "products" },
  { label: "Services", value: "services" },
  { label: "Fees", value: "fees" },
  { label: "Subscriptions/Recurring Charges", value: "subscriptions" },
  { label: "Travel & Lodging", value: "travelLodging" },
  { label: "Marketing & Advertising", value: "marketing" },
  { label: "Office/Administrative", value: "officeAdmin" },
  { label: "Utilities & Overhead", value: "utilities" },
  { label: "Taxes", value: "taxes" },
  { label: "Other", value: "other" },
];

export const GENERIC_FORM_FIELDS = {
  type: "text",
  variant: "outlined",
};

export const TEXTAREA_FORM_FIELDS = {
  multiline: true,
  minRows: 4,
  variant: "outlined",
};

/**
 * Blank Invoice Details form
 */
export const BLANK_INVOICE_DETAILS_FORM = {
  title: {
    id: "title",
    value: "",
    errorMsg: "",
    isRequired: false,
    validators: [],
  },
  caption: {
    id: "caption",
    value: "",
    errorMsg: "",
    isRequired: false,
    validators: [],
  },
  note: {
    id: "note",
    value: "",
    errorMsg: "",
    isRequired: false,
    validators: [
      {
        validate: (value) => value.trim().length >= 150,
        message: "Notes should be less than 150 characters",
      },
    ],
    ...TEXTAREA_FORM_FIELDS,
  },
  start_date: {
    id: "start_date",
    value: dayjs().toISOString(), // default value; prevents leak
    errorMsg: "",
    isRequired: true,
    validators: [
      {
        validate: (value) => value.trim().length <= 0,
        message: "Start Date is required",
      },
      {
        validate: (value) => value.trim().length >= 15,
        message: "Start Date should be less than 15 characters",
      },
    ],
  },
  end_date: {
    id: "end_date",
    value: dayjs().toISOString(), // default value; prevents leak
    errorMsg: "",
    isRequired: true,
    validators: [
      {
        validate: (value) => value.trim().length <= 0,
        message: "End date is required",
      },
      {
        validate: (value) => value.trim().length >= 150,
        message: "End date should be less than 150 characters",
      },
    ],
  },
  tax_rate: {
    id: "tax_rate",
    value: "",
    errorMsg: "",
    isRequired: false,
    validators: [
      {
        validate: (value) => value.trim().length <= 0,
        message: "Tax rate is required",
      },
      {
        validate: (value) => value.trim().length >= 5,
        message: "Tax rate should be less than 5 characters",
      },
      {
        validate: (value) => !/^\d{1,2}(\.\d{1,2})?$/.test(value),
        message: "Tax rate should be a valid tax rate (e.g., 7.25)",
      },
    ],
  },
  invoice_header: {
    id: "invoice_header",
    value: "",
    errorMsg: "",
    isRequired: false,
    validators: [],
  },
};

/**
 * Blank Invoice Details form
 */
export const BLANK_INVOICE_LINE_ITEM_FORM = {
  category: {
    id: "category",
    value: "",
    errorMsg: "",
    isRequired: true,
    validators: [
      {
        validate: (value) => value.length <= 0,
        message: "Select the category of the single item.",
      },
    ],
  },
  descpription: {
    id: "descpription",
    value: "",
    errorMsg: "",
    isRequired: false,
    validators: [],
  },
  caption: {
    id: "caption",
    value: "",
    errorMsg: "",
    isRequired: false,
    validators: [],
  },
  quantity: {
    id: "quantity",
    value: "",
    errorMsg: "",
    isRequired: true,
    validators: [
      {
        validate: (value) => !/^\d+$/.test(value),
        message: "Quantity must be a positive integer",
      },
      {
        validate: (value) => parseInt(value, 10) > 9999,
        message: "Quantity should be less than or equal to 9999",
      },
    ],
  },
  price: {
    id: "price",
    value: "",
    errorMsg: "",
    isRequired: true,
    validators: [
      {
        validate: (value) => !/^\d+(\.\d{1,2})?$/.test(value),
        message: "Price must be a valid number with up to two decimal places",
      },
      {
        validate: (value) => parseFloat(value) <= 0,
        message: "Price must be greater than 0",
      },
    ],
  },
  payment: {
    id: "payment",
    value: "",
    errorMsg: "",
    isRequired: true,
    validators: [
      {
        validate: (value) => !/^\d+(\.\d{1,2})?$/.test(value),
        message: "Payment must be a valid number with up to two decimal places",
      },
    ],
  },
  payment_method: {
    id: "payment_method",
    value: "",
    errorMsg: "",
    isRequired: true,
    validators: [
      {
        validate: (value) => value.trim().length <= 0,
        message: "Method of payment is required",
      },
      {
        validate: (value) => value.trim().length >= 150,
        message: "Method of payment should be less than 150 characters",
      },
    ],
  },
};
