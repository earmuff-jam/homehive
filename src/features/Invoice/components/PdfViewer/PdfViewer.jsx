import React from "react";

import { useNavigate, useOutletContext } from "react-router-dom";

import dayjs from "dayjs";

import { Container, Stack, Typography } from "@mui/material";
import RowHeader from "common/RowHeader/RowHeader";
import Salutation from "features/Invoice/components/UserInfo/Salutation";
import { EditInvoiceRouteUri } from "common/utils";
import EmptyPdfViewer from "features/Invoice/components/PdfViewer/EmptyPdfViewer";
import ReportTable from "features/Invoice/components/PdfViewer/ReportTable";
import { useAppTitle } from "hooks/useAppTitle";

export default function PdfViewer() {
  useAppTitle("View Invoice");

  const navigate = useNavigate();
  const [showWatermark] = useOutletContext();

  const senderInfo = JSON.parse(localStorage.getItem("senderInfo"));
  const recieverInfo = JSON.parse(localStorage.getItem("recieverInfo"));
  const invoice_form = JSON.parse(localStorage.getItem("pdfDetails"));
  const invoiceStatus = JSON.parse(localStorage.getItem("invoiceStatus"));

  const handleNavigate = () => navigate(EditInvoiceRouteUri);

  return (
    <Container maxWidth="md" data-tour="view-pdf-0">
      {!invoice_form ? (
        <EmptyPdfViewer handleNavigate={handleNavigate} />
      ) : (
        <Stack spacing={"2rem"}>
          {recieverInfo ? <Salutation userInfo={recieverInfo} /> : null}
          <RowHeader
            title={invoice_form.title}
            caption={invoice_form.caption}
            showDate={true}
            createdDate={dayjs(invoice_form?.updatedOn?.fromNow).format(
              "DD-MM-YYYY",
            )}
          />
          <Typography
            variant="subtitle2"
            color="text.secondary"
            sx={{ fontStyle: "italic" }}
          >
            Period {invoice_form.start_date} to {invoice_form.end_date}
          </Typography>
          <ReportTable
            rows={invoice_form.items || []}
            taxRate={invoice_form.tax_rate}
            invoiceTitle={invoice_form.invoice_header}
            invoiceStatus={invoiceStatus}
            showWatermark={showWatermark}
          />
          {invoice_form?.note.length > 0 && (
            <Typography
              variant="caption"
              fontStyle="italic"
              fontWeight="medium"
            >
              Note: {invoice_form?.note}
            </Typography>
          )}
          {senderInfo ? (
            <Salutation isEnd={true} userInfo={senderInfo} />
          ) : null}
        </Stack>
      )}
    </Container>
  );
}
