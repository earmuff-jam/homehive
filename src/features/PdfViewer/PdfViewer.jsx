import { Container, Stack, Typography } from "@mui/material";
import ReportTable from "./ReportTable";
import RowHeader from "../../common/RowHeader/RowHeader";
import dayjs from "dayjs";
import { useNavigate } from "react-router-dom";
import Salutation from "../../common/UserInfo/Salutation";

export default function PdfViewer() {
  const navigate = useNavigate();
  const senderInfo = JSON.parse(localStorage.getItem("senderInfo"));
  const recieverInfo = JSON.parse(localStorage.getItem("recieverInfo"));
  const invoice_form = JSON.parse(localStorage.getItem("pdfDetails"));

  const handleNavigate = () => navigate("/edit");

  if (!invoice_form) {
    return (
      <Stack textAlign={"center"}>
        <Typography>Sorry, no invoice found to display</Typography>
        <Typography variant="caption">
          Create new invoice form
          <Typography
            component={"span"}
            variant="caption"
            color="primary"
            sx={{ cursor: "pointer" }}
            onClick={handleNavigate}
          >
            {" "}
            here.
          </Typography>
        </Typography>
      </Stack>
    );
  }

  return (
    <Container maxWidth="md">
      <Stack spacing={"2rem"}>
        {recieverInfo ? <Salutation userInfo={recieverInfo} /> : null}
        <RowHeader
          title={invoice_form.title}
          caption={invoice_form.caption}
          showDate={true}
          createdDate={dayjs(invoice_form.updated_on.fromNow).format(
            "DD-MM-YYYY"
          )}
        />
        <Typography
          variant="subtitle2"
          sx={{ color: "#666", fontStyle: "italic" }}
        >
          Period {invoice_form.start_date} to {invoice_form.end_date}
        </Typography>
        <ReportTable
          rows={invoice_form.items || []}
          taxRate={invoice_form.tax_rate}
          invoiceTitle={invoice_form.invoice_header}
        />
        {invoice_form?.note.length > 0 && (
          <Typography variant="caption" fontStyle="italic" fontWeight="medium">
            Note: {invoice_form?.note}
          </Typography>
        )}
      </Stack>

      {senderInfo ? <Salutation isEnd={true} userInfo={senderInfo} /> : null}
    </Container>
  );
}
