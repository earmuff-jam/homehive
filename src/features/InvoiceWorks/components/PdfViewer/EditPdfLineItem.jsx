import React from "react";

import { Autocomplete, Stack, TextField, Typography } from "@mui/material";
import TextFieldWithLabel from "common/UserInfo/TextFieldWithLabel";
import { InvoiceCategoryOptions } from "features/InvoiceWorks/components/PdfViewer/constants";

export default function EditPdfLineItem({
  index,
  formData,
  handleLineItemChange,
  handleLineItemAutocompleteChange,
}) {
  return (
    <Stack spacing={2}>
      <Stack direction="row" spacing={2} alignItems="center">
        <Stack>
          <Typography variant="body2" fontWeight="medium" gutterBottom>
            Category {formData.category.isRequired && `*`}
          </Typography>
          <Autocomplete
            id="category"
            disablePortal
            options={InvoiceCategoryOptions}
            sx={{ width: 300 }}
            value={formData.category?.selectedOption || ""}
            onChange={(event, newValue) =>
              handleLineItemAutocompleteChange(index, "category", newValue)
            }
            renderInput={(params) => (
              <TextField
                {...params}
                size="small"
                placeholder="Select type of category"
              />
            )}
          />
        </Stack>

        {/* Description */}
        <TextFieldWithLabel
          label="Description"
          id="descpription"
          name="descpription"
          placeholder="Description of charge"
          value={formData?.descpription.value || ""}
          handleChange={(ev) => handleLineItemChange(ev, index)}
          errorMsg={formData.descpription["errorMsg"]}
        />
      </Stack>

      {/* Caption */}
      <TextFieldWithLabel
        label="Caption"
        id="caption"
        name="caption"
        placeholder="Additional details"
        value={formData?.caption.value || ""}
        handleChange={(ev) => handleLineItemChange(ev, index)}
        errorMsg={formData.caption["errorMsg"]}
      />

      {/* Quantity and Price */}
      <Stack direction="row" spacing={2}>
        <TextFieldWithLabel
          label="Quantity *"
          id="quantity"
          name="quantity"
          placeholder="Quantity"
          value={formData?.quantity.value || ""}
          handleChange={(ev) => handleLineItemChange(ev, index)}
          errorMsg={formData.quantity["errorMsg"]}
        />
        <TextFieldWithLabel
          label="Price *"
          id="price"
          name="price"
          placeholder="Total cost of current item in USD"
          value={formData?.price.value || ""}
          handleChange={(ev) => handleLineItemChange(ev, index)}
          errorMsg={formData.price["errorMsg"]}
        />
      </Stack>

      {/* Payment and Mode of Payment */}
      <Stack direction="row" spacing={2}>
        <TextFieldWithLabel
          label="Payment *"
          id="payment"
          name="payment"
          placeholder="Amount paid in USD."
          value={formData?.payment.value || ""}
          handleChange={(ev) => handleLineItemChange(ev, index)}
          errorMsg={formData.payment["errorMsg"]}
        />
        <TextFieldWithLabel
          label="Mode of Payment*"
          id="payment_method"
          name="payment_method"
          placeholder="Payment Method, eg., Zelle, Money Order"
          value={formData?.payment_method.value || ""}
          handleChange={(ev) => handleLineItemChange(ev, index)}
          errorMsg={formData.payment_method["errorMsg"]}
        />
      </Stack>
    </Stack>
  );
}
