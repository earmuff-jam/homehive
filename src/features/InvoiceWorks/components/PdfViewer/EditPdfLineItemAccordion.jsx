import { DeleteRounded, ExpandMoreRounded } from "@mui/icons-material";
import {
  Accordion,
  AccordionDetails,
  AccordionSummary,
  IconButton,
  Stack,
  Typography,
} from "@mui/material";
import EditPdfLineItem from "features/InvoiceWorks/components/PdfViewer/EditPdfLineItem";

export default function EditPdfLineItemAccordion({
  lineItem,
  title,
  index,
  handleDelete,
  handleLineItemChange,
  handleLineItemAutocompleteChange,
}) {
  return (
    <Accordion defaultExpanded>
      <AccordionSummary
        expandIcon={<ExpandMoreRounded />}
        aria-controls="panel1-content"
        id="panel1-header"
      >
        <Stack alignItems={"center"} direction="row">
          <IconButton size="small" onClick={() => handleDelete(index)}>
            <DeleteRounded fontSize="small" color="error" />
          </IconButton>
          <Typography>{title}</Typography>
        </Stack>
      </AccordionSummary>
      <AccordionDetails>
        <EditPdfLineItem
          index={index}
          formData={lineItem}
          handleLineItemChange={handleLineItemChange}
          handleLineItemAutocompleteChange={handleLineItemAutocompleteChange}
        />
      </AccordionDetails>
    </Accordion>
  );
}
