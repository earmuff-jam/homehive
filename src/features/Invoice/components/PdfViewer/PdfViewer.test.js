import React from "react";
import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import PdfViewer from "./PdfViewer";


jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => jest.fn(),
  useOutletContext: () => [false], // mock value for showWatermark
}));

test("PdfViewer snapshot", () => {
  const { asFragment } = render(
    <MemoryRouter>
      <PdfViewer />
    </MemoryRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});
