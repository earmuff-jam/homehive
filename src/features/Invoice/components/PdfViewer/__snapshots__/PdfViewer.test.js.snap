// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PdfViewer snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiContainer-root MuiContainer-maxWidthMd css-u4uvlt-MuiContainer-root"
    data-tour="view-pdf-0"
  >
    <div
      class="MuiStack-root css-1ug9gnt-MuiStack-root"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
      >
        Sorry, no invoice found to display
      </p>
      <div
        class="MuiStack-root css-nen11g-MuiStack-root"
      >
        <span
          class="MuiTypography-root MuiTypography-caption css-mpaqru-MuiTypography-root"
        >
          Create new invoice from 
          <span
            class="MuiTypography-root MuiTypography-caption css-mf4qso-MuiTypography-root"
          >
            Â here.
          </span>
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
